File: 02/one.rb
# frozen_string_literal: true

require 'pathname'
require 'pry-nav'

arg = ARGV.shift

case arg
when /input|input.txt/
  FILE_PATH = Pathname.new(File.expand_path('input.txt'))
when /test|test.txt/
  FILE_PATH = Pathname.new(File.expand_path('test.txt'))
else
  raise ArgumentError, "Expects 'input' or 'test' as command line argument"
end

lines = File.open(FILE_PATH, File::RDONLY).readlines(chomp: true)

def increasing?(vals)
  prev = vals.first
  (1...vals.length).each do |i|
    cur = vals[i]
    return false if cur <= prev

    diff = (cur - prev).abs
    return false if diff < 1 || diff > 3
  
    prev = cur
  end
  true
end

def decreasing?(vals)
  prev = vals.first
  (1...vals.length).each do |i|
    cur = vals[i]
    return false if cur >= prev

    diff = (cur - prev).abs
    return false if diff < 1 || diff > 3

    prev = cur
  end
  true
end

def safe?(line)
  vals = line.split(/\s+/).map(&:to_i)
  diff = vals[0] - vals[1]
  return false if diff == 0

  if diff < 0
    return increasing?(vals)
  else
    return decreasing?(vals)
  end
end

res = 0
lines.each do |line|
  res += 1 if safe?(line)
end
puts res

# answer: 483

File: 02/one.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"regexp"
	"strconv"
)

func main() {
	args := os.Args[1:]
	if len(args) != 1 {
		panic("error: expects 'input' or 'test' as an argument")
	}

	var filepath string
	arg := args[0]
	if matched, err := regexp.Match("test", []byte(arg)); matched {
		if err != nil {
			panic(err)
		}
		filepath = "test.txt"
	} else if matched, err := regexp.Match("input", []byte(arg)); matched {
		if err != nil {
			panic(err)
		}
		filepath = "input.txt"
	} else {
		panic("error: expects 'input' or 'test' as an argument")
	}

	file, err := os.Open(filepath)
	if err != nil {
		panic(err)
	}

	scanner := bufio.NewScanner(file)
	lines := []string{}
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	res := 0
	for _, line := range lines {
		if isSafe(line) {
			res++
		}
	}
	fmt.Println(res)
}

func isSafe(line string) bool {
	vals := toi(line)
	diff := vals[0] - vals[1]
	if diff == 0 {
		return false
	}
	if diff < 0 {
		return isIncreasing(vals)
	} else {
		return isDecreasing(vals)
	}
}

func toi(line string) []int {
	vals := []int{}
	split := regexp.MustCompile(`\s+`).Split(line, -1)
	for _, char := range split {
		num, err := strconv.Atoi(char)
		if err != nil {
			panic(err)
		}
		vals = append(vals, num)
	}
	return vals
}

func isIncreasing(vals []int) bool {
	prev := vals[0]
	for i := 1; i < len(vals); i++ {
		cur := vals[i]
		if cur <= prev {
			return false
		}
		diff := int(math.Abs(float64(cur - prev)))
		if diff < 1 || diff > 3 {
			return false
		}
		prev = cur
	}
	return true
}

func isDecreasing(vals []int) bool {
	prev := vals[0]
	for i := 1; i < len(vals); i++ {
		cur := vals[i]
		if cur >= prev {
			return false
		}
		diff := int(math.Abs(float64(cur - prev)))
		if diff < 1 || diff > 3 {
			return false
		}
		prev = cur
	}
	return true
}

// answer: 483
